Revise docs for behavior change flags (#5389)
fishtown-analytics/docs.getdbt.com: website/docs/reference/global-configs/legacy-behaviors.md

- Providing maintainability of dbt software. Every fork in behavior requires additional testing & cognitive overhead that slows future development. These flags exist to facilitate migration from "current" to "better," not to stick around forever.

These flags go through three phases of development:

> 1. **Introduction (Disabled by Default):** The logic is added within dbt to support both 'old' + 'new' behaviours. The 'new' behavior is gated by behind a flag, which is disabled by default, preserving the old behavior.

Reviewer: ```suggestion
1. **Introduction (disabled by default):** dbt adds logic to support both 'old' + 'new' behaviors. The 'new' behavior is gated behind a flag, disabled by default, preserving the old behavior.
```

> 2. **Maturity (Enabled by Default):** The default value of the flag is switched, from `false` to `true`, enabling the new behavior by default. Users may still opt out of the 'new' behavior, preserving the 'old' behavior, by setting the flag to `false` in their projects. They may see deprecation warnings when they do so.

Reviewer: ```suggestion
2. **Maturity (enabled by default):** The default value of the flag is switched, from `false` to `true`, enabling the new behavior by default. Users can preserve the 'old' behavior and opt out of the 'new' behavior by setting the flag to `false` in their projects. They may see deprecation warnings when they do so.
```

> 3. **Removal (Generally Enabled):** The flag, having been marked in advance for deprecation, is removed. The logic to support the 'old' behaviour is removed from dbt codebases. Our intent is to support most flags indefinitely, but we are not committed to supporting them forever. If we do choose to remove a flag, we will offer significant advance notice.

Reviewer: ```suggestion
3. **Removal (generally enabled):** After marking the flag for deprecation, we remove it along with the 'old' behavior it supported from the dbt codebases. We aim to support most flags indefinitely, but we're not committed to supporting them forever. If we choose to remove a flag, we'll offer significant advance notice.
```

> ## Behavior Change Flags

Reviewer: ```suggestion
## Behavior change flags
```

> These flags can _only_ be set in the `flags` dictionary in `dbt_project.yml`. They configure behaviors closely tied to project code, and as such they should be defined in version control and modified via pull/merge request, with the same testing and peer review.

Reviewer: ```suggestion
These flags _must_ be set in the `flags` dictionary in `dbt_project.yml`. They configure behaviors closely tied to project code, which means they should be defined in version control and modified through pull or merge requests, with the same testing and peer review.
```

> The following example displays the current flags and their current default values in the latest dbt Cloud and dbt Core versions. To opt out of a specific behavior change, set the values of the flag to `False` in `dbt_project.yml`. You'll continue to see warnings for legacy behaviors that you have opted out of explicitly until you either resolve them (switch the flag to `True`) or choose to silence the warnings.

Reviewer: @jtcohen6 for teh warm error options - it doesn't look like we have a `.silence` mentioned on the [page](https://docs.getdbt.com/reference/global-configs/warnings) - should we add that so that we can link it in this sentence?

```suggestion
The following example displays the current flags and their current default values in the latest dbt Cloud and dbt Core versions. To opt out of a specific behavior change, set the values of the flag to `False` in `dbt_project.yml`. You'll continue to see warnings for legacy behaviors that you have opted out of explicitly until you either resolve them (switch the flag to `True`) or choose to silence the warnings using the `warn_error_options.silence` flag. 
```

Author: This was _just_ added:
- https://github.com/dbt-labs/docs.getdbt.com/issues/5417

Reviewer 2: I added Mrina's suggestion and we can link to the docs once it's documented!

> The current flags are shown below, along with their current default values in the latest versions of dbt Cloud + dbt Core. To opt out of a specific behavior change, you should set the values of the flag to `False` in `dbt_project.yml`. You will continue to see warnings for legacy behaviors that you have opted out of explicitly, until you either resolve them (switch the flag to `True`) or choose to silence the warnings.

Reviewer: ```suggestion
The following example displays the current flags and their current default values in the latest dbt Cloud and dbt Core versions. To opt out of a specific behavior change, set the values of the flag to `False` in `dbt_project.yml`. You'll continue to see warnings for legacy behaviors that you have opted out of explicitly until you either resolve them (switch the flag to `True`) or choose to silence the warnings.
```

<File name='dbt_project.yml'>

```yml
flags:
  require_explicit_package_overrides_for_builtin_materializations: False
  require_model_names_without_spaces: False
  source_freshness_run_project_hooks: False
```

</File>

| Flag                                                            | dbt Cloud: Intro | dbt Cloud: Maturity | dbt Core: Intro | dbt Core: Maturity | 
|-----------------------------------------------------------------|------------------|---------------------|-----------------|--------------------|
| require_explicit_package_overrides_for_builtin_materializations | 2024.04.141      | 2024.05.xxx         | 1.6.14, 1.7.14    | 1.8.0             |

> | require_resource_names_without_spaces                           | 2024.05.xxx      | 2024.06.xxx         | 1.8.0           | 1.9.0             |

Reviewer: apologies but what does the xx mean?

Author: It means, I don't know what the exact release # is going to be, but the release is planned for `2024.05` (May) or `2024.06` (June). Definitely open to your thoughts — maybe `2024.05.TBD`, or just `2024.05` for now?

https://docs.getdbt.com/reference/commands/version#versioning

<img width="834" alt="Screenshot 2024-05-01 at 21 38 14" src="https://github.com/dbt-labs/docs.getdbt.com/assets/13897643/01ac126b-1f73-40c1-8fac-1e0096be45fc">

Reviewer 2: Changed to TBD

| source_freshness_run_project_hooks                              | 2024.03.61       | 2024.06.xxx         | 1.8.0           | 1.9.0             |

> <sup>1</sup>dbt Cloud - "Keep on latest version"

Reviewer: what is the '1' referencing?

Author: Nothing - I previously had this as a footnote in the table, but then I think I dropped it.

Idea here is just, any time we say "dbt Cloud" in this doc / in the table above, we're talking about "Keep on latest version"

Reviewer 2: Moved this info above table

> ### `require_explicit_package_overrides_for_builtin_materializations`

Reviewer: for the headers, i'm not sure theyre easy to read in the right toc. wondering if we can make the headers be shorter and more direct while including a brief description of the flag right within the header itself?
![Screenshot 2024-05-01 at 12 43 15](https://github.com/dbt-labs/docs.getdbt.com/assets/89008547/46db934a-c0d1-4fd5-abc9-6cbc6268da51)

Reviewer: if youre ok with it, i can update the links in the release notes and guides!

Author: Agree, they look awkward in the sidebar. Without headers is fine. We have links in `dbt-core` pointing to this page with anchors, it's no big deal if the anchors don't resolve correctly (the page load still works fine), I'll just update those links for v1.8 / "KoLV" once we've merged this PR.

> ### `require_explicit_package_overrides_for_builtin_materializations`

Reviewer: ```suggestion
###  Package override for built-in materialization 
```

> We have deprecated the behavior whereby installed packages could override built-in materializations without explicit opt-in from the end user. When this flag is set to `True`, a materialization defined in a package that matches the name of a built-in materialization will no longer be included in the search and resolution order. Unlike for macros, materializations do not use the `search_order` defined in the project `dispatch` config.

Reviewer: so if we update the header, we can have a mention of the actual flag in the first few sentences, like so:

```suggestion
Setting the `require_explicit_package_overrides_for_builtin_materializations` flag to `True` prevents this automatic override. 

We have deprecated the behavior where installed packages could override built-in materializations without your explicit opt-in. When this flag is set to `True`, a materialization defined in a package that matches the name of a built-in materialization will no longer be included in the search and resolution order. Unlike macros, materializations don't use the `search_order` defined in the project `dispatch` config.
```

> The built-in materializations are `'view'`, `'table'`, `'incremental'`, `'materialized_view'` for models as well as `'test'`, `'unit'`, `'snapshot'`, `'seed'`, and `'clone'`.

Reviewer: this makes it seem like test, unit, snapshots, are also materializations?  do we mean to say 'the builtin materializations for models are xyz. The builtin dbt resources (?)  are tests, seed, etc.

```suggestion
The built-in [materializations](/docs/build/materializations) are `'view'`, `'table'`, `'incremental'`, `'materialized_view'` for models as well as `'test'`, `'unit'`, `'snapshot'`, `'seed'`, and `'clone'`.
```

Author: Fun bit of trivia! :)

Technically "seed" is a builtin resource type in dbt, and seeds use a built-in _materialization_ that's also called `seed`. The same goes for tests, unit tests, and snapshots. Users actually could (if they really wanted to) modify those materializations. The same was also true of third-party package maintainers, and that's the behavior we're modifying here.

<img width="846" alt="Screenshot 2024-05-01 at 21 45 01" src="https://github.com/dbt-labs/docs.getdbt.com/assets/13897643/26e5e4fe-cd94-46fd-974b-4bb3660664e7">

Reviewer 2: @mirnawong1 Closing this as resolved, but let me know if anythign needs to happen here!

Reviewer: oh that's interesting! thanks for sharing @jtcohen6 ! and all good, ty @runleonarun !

> Users can still explicitly override built-in materializations, in favor of a materialization defined in a package, by reimplementing the built-in materialization in their root project and wrapping the package implementation.

Reviewer: ```suggestion
You can still explicitly override built-in materializations, in favor of a materialization defined in a package, by reimplementing the built-in materialization in your root project and wrapping the package implementation.
```

<File name='macros/materialization_view.sql'>

```sql
{% materialization view, snowflake %}
  {{ return(my_installed_package_name.materialization_view_snowflake()) }}
{% endmaterialization %}
```

</File>

> In the future, we may extend the project-level `dispatch` configuration [TODO: link] to support a list of authorized packages for overriding built-in materialization.

Reviewer: ```suggestion
In the future, we may extend the project-level [`dispatch` configuration](/reference/project-configs/dispatch-config) to support a list of authorized packages for overriding built-in materialization.
```

Reviewer: adding link

<VersionBlock lastVersion="1.7">

> The flags below have been introduced in a future version of dbt Core. If you're still using an older version, then you have the legacy behavior (when each flag is `False`).

Reviewer: ```suggestion
The following flags were introduced in a future version of dbt Core. If you're still using an older version, then you have the legacy behavior by default (when each flag is `False`). 
```

</VersionBlock>

> ### `require_resource_names_without_spaces`

Reviewer: ```suggestion
### No spaces in resource names
```

> The names of dbt resources (models, sources, etc) should contain letters, numbers, and underscores. We highly discourage the use of other characters, especially spaces. To that end, we have deprecated support for spaces in resource names. When this flag is set to the `True`, dbt will raise an exception (instead of a deprecation warning) if it detects a space in a resource name.

Reviewer: ```suggestion
The `require_resource_names_without_spaces` flag enforces using resource names without spaces. 

The names of dbt resources (models, sources, etc) should contain letters, numbers, and underscores. We highly discourage the use of other characters, especially spaces. To that end, we have deprecated support for spaces in resource names. When the `require_resource_names_without_spaces` flag is set to `True`, dbt will raise an exception (instead of a deprecation warning) if it detects a space in a resource name.
```

<File name='models/model name with spaces.sql'>

```sql
-- This model file should be renamed to model_name_with_underscores.sql
```

</File>

> ### `source_freshness_run_project_hooks`

Reviewer: ```suggestion
### Project hooks with source freshness 
```

> This configures the `dbt source freshness` command to include "project hooks" (`on-run-start` / `on-run-end`, TODO add link) as part of their execution.

Reviewer: ```suggestion
Set the `source_freshness_run_project_hooks` flag to `True` to include "project hooks" ([`on-run-start` / `on-run-end`](/reference/project-configs/on-run-start-on-run-end)) in the `dbt source freshness` command execution.
```
